tools/c7n_azure/c7n_azure/resources/key_vault.py

```diff
@@ -23,3 +30,69 @@ class resource_type(ArmResourceManager.resource_type):
         service = 'azure.mgmt.keyvault'
         client = 'KeyVaultManagementClient'
         enum_spec = ('vaults', 'list', None)
+
+
+@KeyVault.filter_registry.register('whitelist')
+class WhiteListFilter(ValueFilter):
+    schema = type_schema('whitelist', rinherit=ValueFilter.schema)
+    graph_client = None
+
+    def __call__(self, i):
+        if 'access_policies' not in i:
+            client = self.manager.get_client()
+            vault = client.vaults.get(i['resourceGroup'], i['name'])
+
+            # Retrieve access policies for the keyvaults
+            access_policies = []
+            for policy in vault.properties.access_policies:
+                access_policies.append({
+                    'tenant_id': policy.tenant_id,
+                    'object_id': policy.object_id,
+                    'application_id': policy.application_id,
+                    'permissions': {
+                        'keys': policy.permissions.keys,
+                        'secrets': policy.permissions.secrets,
+                        'certificates': policy.permissions.certificates
+                    }
+                })
+
+            # Enhance access policies with display_name, object_type and principal_name before continue
+            i['access_policies'] = self.enhance_policies(access_policies)
+
+        return super(WhiteListFilter, self).__call__(i)
+
+    def enhance_policies(self, access_policies):
+        if self.graph_client is None:
```

Comment:
So permissions to talk to the graph are not guaranteed with all the possible types of authorization we provide. For example, in our tenant an SP will never have access to talk to the graph because a tenant admin needs to approve it. You can only access the graph through cli auth locally. We'd need a try catch on any of the graph calls or this will throw an error.

Reply:
You actually might be able to use a related resource filter instead. We've used it in a couple of spots, but essentially it's a join on another resource type by a key. So if the role_assignment has the principalId that relates to a key on the key vault policies then you could filter through a related filter instead.

Reply:
We can't use related filter here because SPs with a KV access might be in the different subscriptions within the same tenant.